heron/schedulers/src/java/org/apache/heron/scheduler/kubernetes/KubernetesContext.java

```diff
@@ -291,31 +309,202 @@ public static boolean getPersistentVolumeClaimDisabled(Config config) {
+   */
+  public static Map<String, Map<KubernetesConstants.VolumeConfigKeys, String>>
+      getVolumeClaimTemplates(Config config, boolean isExecutor) {
+    final Matcher matcher = KubernetesConstants.VALID_LOWERCASE_RFC_1123_REGEX.matcher("");
+
+    final Map<String, Map<KubernetesConstants.VolumeConfigKeys, String>> volumes =
+        getVolumeConfigs(config, KubernetesContext.KUBERNETES_VOLUME_CLAIM_PREFIX, isExecutor);
+
+    for (Map.Entry<String, Map<KubernetesConstants.VolumeConfigKeys, String>> volume
+        : volumes.entrySet()) {
+
+      // Claim name is required.
+      if (!volume.getValue().containsKey(KubernetesConstants.VolumeConfigKeys.claimName)) {
+        throw new TopologySubmissionException(String.format("Volume `%s`: Persistent Volume"
+            + " Claims require a `claimName`.", volume.getKey()));
+      }
+
+      for (Map.Entry<KubernetesConstants.VolumeConfigKeys, String> volumeConfig
+          : volume.getValue().entrySet()) {
+        final KubernetesConstants.VolumeConfigKeys key = volumeConfig.getKey();
+        final String value = volumeConfig.getValue();
+
+        switch (key) {
+          case claimName:
+            // Claim names which are not OnDemand should be lowercase RFC-1123.
+            if (!matcher.reset(value).matches()
+                && !KubernetesConstants.LABEL_ON_DEMAND.equalsIgnoreCase(value)) {
+              throw new TopologySubmissionException(String.format("Volume `%s`: `claimName` does"
+                  + " not match lowercase RFC-1123 pattern", volume.getKey()));
+            }
+            break;
+          case storageClassName:
+            if (!matcher.reset(value).matches()) {
+              throw new TopologySubmissionException(String.format("Volume `%s`: `storageClassName`"
+                  + " does not match lowercase RFC-1123 pattern", volume.getKey()));
+            }
+            break;
+          case sizeLimit: case accessModes: case volumeMode: case readOnly: case path: case subPath:
+            break;
+          default:
+            throw new TopologySubmissionException(String.format("Volume `%s`: Invalid Persistent"
+                + " Volume Claim type option for '%s'", volume.getKey(), key));
+        }
+      }
+    }
+
+    return volumes;
+  }
+
+  /**
+   * Collects parameters form the <code>CLI</code> and validates options for <code>Empty Directory</code>s.
+   * @param config Contains the configuration options collected from the <code>CLI</code>.
+   * @param isExecutor Flag used to collect CLI commands for the <code>Executor</code> and <code>Manager</code>.
+   * @return A mapping between <code>Volumes</code> and their configuration <code>key-value</code> pairs.
+   * Will return an empty list if there are no Volume Claim Templates to be generated.
+   */
+  public static Map<String, Map<KubernetesConstants.VolumeConfigKeys, String>>
+      getVolumeEmptyDir(Config config, boolean isExecutor) {
```

Comment:
Would it be worthwhile to mark this parameters passed in the function as `final`?

Reply:
Changes affected.